proxy/__init__.py

```diff
@@ -0,0 +1,116 @@
+Registries following the distribution spec are supported.
+"""
+from __future__ import annotations
+from typing import Any
+from urllib.parse import urlparse
+import re
+
+import requests
+
+
+# BLOB_CONTENT_TYPE = "application/octet-stream"
+DEFAULT_SCHEME = "https"
+WWW_AUTHENTICATE_REGEX = re.compile(r'(\w+)[=] ?"?([^",]+)"?')
+
+
+def parse_www_auth(value: str) -> dict[str, str]:
+    """
+    Parses WWW-Authenticate parameters and returns a dict of key=val.
+    See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+    for details.
+    """
+    matches = WWW_AUTHENTICATE_REGEX.findall(value)
+    return dict(matches)
+
+
+class Proxy:
+    def __init__(self, registry: str, repository: str, auth: tuple[str, str] | None = None):
+        url = registry.rstrip("/")
+        if not registry.startswith("http"):
+            url = f"{DEFAULT_SCHEME}://{registry}"
+        self.base_url = url
+        self.session = requests.Session()
+        self._repo = repository
+        self._authorize(auth)
+
+    def get_manifest(self, image_ref: str, media_type: str | None = None) -> dict[str, Any]:
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+
+        url = f"{self.base_url}/v2/{self._repo}/manifests/{image_ref}"
+        resp = self.session.get(
+            url,
+            headers=headers,
+        )
+        return {
+            "response": resp.text,
+            "status": resp.status_code,
+            "headers": dict(resp.headers),
+        }
+
+    def get_blob(self, digest: str, media_type: str | None = None):
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+
+        url = f"{self.base_url}/v2/{self._repo}/blobs/{digest}"
+        resp = self.session.get(
+            url,
+            headers=headers,
+            allow_redirects=True,
+            stream=True,
+        )
+        if not resp.ok:
+            return {
+                "response": resp.text,
+                "status": resp.status_code,
+                "headers": dict(resp.headers),
+            }
+
+        def stream_contents():
+            chunk = 1024
+            for chunk in resp.iter_content(chunk_size=chunk):
+                yield chunk
+
+        # blob_size = int(resp.headers.get("Content-Length"))
+        # headers = {
+        #     "Docker-Content-Digest": digest,
+        #     "Content-Type": BLOB_CONTENT_TYPE,
+        #     "Content-Length": blob_size,
+        # }
+        # accept_ranges = resp.headers.get("Accept-Ranges", None)
+        # if accept_ranges is not None:
+        #     headers["Accept-Ranges"] = accept_ranges
+        return {
+            "response": stream_contents(),
+            "status": resp.status_code,
+            "headers": dict(resp.headers),
+        }
+
+    def _authorize(self, auth: tuple[str, str] | None = None):
```

Comment:
@flavianmissi does it work for registries that use basic auth?

Reply:
Not with the code as is. Are there any such registries? 🤔

Reply:
Should be easy enough to support it though, we just need the `parse_www_auth` to return the auth scheme, then use when adding the authentication header to the session.

Reply:
there are, but they aren't public

Reply:
@dmesser do we care about registries that use basic auth? an option would be to list this as a known limitation for now, then if anyone needs it it's easy enough to add it.

Reply:
basic auth works. kudos to @dmage for helping me test it 🚀